src/robusta/integrations/resource_analysis/memory_analyzer.py

```diff
@@ -0,0 +1,125 @@
+        """
+        Returns the maximal memory usage (in bytes) for the given container, in the time range between now and now - duration.r
+        """
+
+        # We use the container_memory_usage_bytes metric, with the following filters:
+        # - node={node_name}:
+        #   In order to filter on the given node.
+        #
+        # - pod={pod_name}:
+        #   In order to filter on the given pod.
+        #
+        # - image!="":
+        #   In order to ignore the metric of the pod in general, which sums the memory of each of its containers.
+        #
+        # - container!="POD":
+        #   In order to ignore paused containers. For more information, please see:
+        #   https://stackoverflow.com/questions/68683403/what-is-the-container-pod-label-in-prometheus-and-why-do-most-examples-exclude
+        #
+        # - container={container_name}:
+        #   In order to filter on the given container name.
+        #
+        # - id=~"/kubepods/.*":
+        #   In order to filter multiple ids that are returned for the same container.
+        #   In my minikube, for example, 2 time series are returned for each container, one with id=/docker/* and one with id=/kubepods/*.
+        #   However, an id of the form /kubepods/* always exists. Also, the following prefixes are always mutual exclusive:
+        #       /kubepods/burstable/*
+        #       /kubepods/besteffort/*
+        #       /kubepods/guaranteed/*
+        #   Therefore, there will always be exactly one id of the form /kubpods/* prefix for each container.
+        #   See https://stackoverflow.com/questions/49035724/how-do-i-resolve-kubepods-besteffort-poduuid-to-a-pod-name for more information.
+
+        return self._get_max_value_in_first_series_of_query(
+            f"container_memory_usage_bytes{{node=\"{node_name}\", pod=\"{pod_name}\", image!=\"\", "
+            f"container!=\"POD\", container=\"{container_name}\", id=~\"/kubepods/.*\"}}",
+            duration
+        )
+
+    def _get_max_value_in_first_series_of_query(self, promql_query: str, duration: timedelta) -> Optional[float]:
+        results = self._query(promql_query, duration)
+
+        if len(results) == 0:
+            return None
+
+        series = results[0]
+        series_values = series["values"]
+        max_value_in_series = max([float(val) for (timestamp, val) in series_values])
+        return max_value_in_series
+
+    def _query(self, promql_query: str, duration: timedelta) -> list:
+        end_time = datetime.now(tz=self.prometheus_tzinfo)
+        start_time = end_time - duration
+        step = 1
+        results = self.prom.custom_query_range(
+            promql_query,
+            start_time,
+            end_time,
+            step,
+            {
+                "timeout": self.default_prometheus_params["timeout"]
+            }
+        )
+
+        return results
+
+
+class K8sMemoryTransformer:
```

Comment:
This can probably be a static class, it doesn't have any state

Reply:
Changed (initialization of factors is now at the module level, as there are no static initializers in Python)